recipes/LibriSpeech/ASR/CTC/train_with_wav2vec_k2.py

```diff
@@ -0,0 +1,656 @@
+ * Georgios Karakasidis 2023
+ * Rudolf A Braun 2022
+ * Titouan Parcollet 2022
+ * Sung-Lin Yeh 2021
+ * Ju-Chieh Chou 2020
+ * Mirco Ravanelli 2020
+ * Abdel Heba 2020
+ * Peter Plantinga 2020
+ * Samuele Cornell 2020
+"""
+
+import os
+import sys
+from typing import List, Union
+import torch
+import logging
+import speechbrain as sb
+from speechbrain.utils.distributed import run_on_main, if_main_process
+from hyperpyyaml import load_hyperpyyaml
+from pathlib import Path
+
+from speechbrain.k2_integration.prepare_lang import prepare_lang
+from speechbrain.k2_integration.graph_compiler import CtcTrainingGraphCompiler
+from speechbrain.k2_integration.lexicon import Lexicon, get_lexicon
+
+logger = logging.getLogger(__name__)
+
+
+# Define training procedure
+class ASR(sb.Brain):
+    def compute_forward(self, batch, stage):
+        """Forward computations from the waveform batches to the output probabilities."""
+        batch = batch.to(self.device)
+        wavs, wav_lens = batch.sig
+        wavs, wav_lens = wavs.to(self.device), wav_lens.to(self.device)
+        # Downsample the inputs if specified
+        if hasattr(self.modules, "downsampler"):
+            wavs = self.modules.downsampler(wavs)
+        # Add augmentation if specified
+        if stage == sb.Stage.TRAIN:
+            if hasattr(self.hparams, "augmentation"):
+                wavs = self.hparams.augmentation(wavs, wav_lens)
+
+        # Forward pass
+
+        # Handling SpeechBrain vs HuggingFance pretrained models
+        if hasattr(self.modules, "extractor"):  # SpeechBrain pretrained model
+            latents = self.modules.extractor(wavs)
+            feats = self.modules.encoder_wrapper(latents, wav_lens=wav_lens)[
+                "embeddings"
+            ]
+        else:  # HuggingFace pretrained model
+            feats = self.modules.wav2vec2(wavs, wav_lens)
+
+        x = self.modules.enc(feats)
+
+        # Compute outputs
+        logits = self.modules.ctc_lin(x)
+
+        # Upsample the inputs if they have been highly downsampled
+        if hasattr(self.hparams, "upsampling") and self.hparams.upsampling:
+            logits = logits.view(
+                logits.shape[0], -1, self.hparams.output_neurons
+            )
+
+        p_ctc = self.hparams.log_softmax(logits)
+        return p_ctc, wav_lens
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss (CTC+NLL) given predictions and targets."""
+
+        p_ctc, wav_lens = predictions
+
+        ids = batch.id
+
+        # Sort batch to be descending by length of wav files, which is demanded by k2
+        if self.hparams.sorting == "ascending":
+            p_ctc = torch.flip(p_ctc, (0,))
+            wav_lens = torch.flip(wav_lens, (0,))
+            texts = [batch.wrd[i] for i in reversed(range(len(batch.wrd)))]
+        elif self.hparams.sorting == "descending":
+            texts = batch.wrd
+        else:
+            raise NotImplementedError(
```

Comment:
Does this mean that this recipe does not support dynamic batching? Could our users easily extend this work to something that uses dynamic batching? This is the standard way of training these days.

Reply:
We could do something like this to avoid sorting in `compute_objectives` (this could remove the commented if-else block): https://github.com/geoph9/speechbrain/blob/k2-lfmmi/speechbrain/k2_integration/losses.py#L30C3-L48.

I think that would also work for dynamic batching.

Reply:
This recipe is not dynamic batching compliant as we are not using it contrary to other recipes like : `https://github.com/speechbrain/speechbrain/blob/develop/recipes/LibriSpeech/ASR/transducer/train.py#L390`. However, I will give it a try by adding dynamic batching support in local.